buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -96,25 +100,56 @@ final class AdaptivePoolingAllocator {
      */
     private static final int CENTRAL_QUEUE_CAPACITY = SystemPropertyUtil.getInt(
             "io.netty.allocator.centralQueueCapacity", NettyRuntime.availableProcessors());
+    private static final Magazine[] EMPTY_MAGAZINES = new Magazine[0];
 
     private final ChunkAllocator chunkAllocator;
     private final Queue<ChunkByteBuf> centralQueue;
     private final StampedLock magazineExpandLock;
     private volatile Magazine[] magazines;
+    private final CopyOnWriteArraySet<Magazine> liveFastThreadLocalMagazines;
+    private final FastThreadLocal<Magazine> fastThreadLocalMagazines;
 
     AdaptivePoolingAllocator(ChunkAllocator chunkAllocator) {
+        this(chunkAllocator, false);
+    }
+
+    AdaptivePoolingAllocator(ChunkAllocator chunkAllocator, boolean useFastThreadLocalMagazines) {
         this.chunkAllocator = chunkAllocator;
         if (javaVersion() < 8) {
             // The implementation uses StampedLock, which was introduced in Java 8.
             throw new IllegalStateException("This allocator require Java 8 or newer.");
         }
         centralQueue = ObjectUtil.checkNotNull(createSharedChunkQueue(), "centralQueue");
         magazineExpandLock = new StampedLock();
-        Magazine[] mags = new Magazine[4];
-        for (int i = 0; i < mags.length; i++) {
-            mags[i] = new Magazine(this);
+        if (useFastThreadLocalMagazines) {
+            final CopyOnWriteArraySet<Magazine> liveMagazines = new CopyOnWriteArraySet<>();
+            fastThreadLocalMagazines = new FastThreadLocal<Magazine>() {
+                @Override
+                protected Magazine initialValue() {
+                    Magazine magazine = new Magazine(AdaptivePoolingAllocator.this, true);
+                    liveMagazines.add(magazine);
+                    return magazine;
+                }
+
+                @Override
+                protected void onRemoval(Magazine value) {
+                    // TODO let's release in the global pool?
+                    //      or mark the magazine as disposable and defer the release?
+                    liveMagazines.remove(value);
+                }
+            };
+            liveFastThreadLocalMagazines = liveMagazines;
+            // let's allocate these lazily here, 'cause we are optimizing for event loop fast thread local threads!
+            magazines = EMPTY_MAGAZINES;
+        } else {
+            fastThreadLocalMagazines = null;
+            liveFastThreadLocalMagazines = null;
+            Magazine[] mags = new Magazine[INITIAL_MAGAZINES];
+            for (int i = 0; i < mags.length; i++) {
+                mags[i] = new Magazine(this);
+            }
+            magazines = mags;
```

Comment:
Unused magazines take up very little memory. Let's just always initialize them like before.
The `useFastThreadLocalMagazines` should only enable a short-cut path in `allocate()` and determine the allocation of the thread-local.